[FLIRT] Create signatures from RzAnalysis data

wargio opened the pull request:
# squash me

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

1. Implements the creation of FLIRT signatures in `.sig` (compressed) and `.pac` human readable formats
2. Provides customization via `e` 

```
 flirt.node.optimize: FLIRT optimization option when creating a signature file (none: 0, normal: 1, smallest: 2)
   flirt.sig.deflate: enables/disables FLIRT zlib compression when creating a signature file (available only for .sig files)
      flirt.sig.file: FLIRT file list (comma separated) for sig format
   flirt.sig.library: FLIRT library name for sig format
        flirt.sig.os: FLIRT operating system list (comma separated) for sig format
   flirt.sig.version: FLIRT version for sig format
```

3. Implements FLIRT creation via new command `zfc`: example `zfc /my/flirt.pac` or `zfc /my/flirt.sig`
4. Supports `zlib` compression on creation
5. Rewritten rz-sign to support creation, dumping and conversion of FLIRT files and dropped zignatures.

```
Usage: rz-sign [options] [file]
 -h                          this help message
 -a [-a]                     add extra 'a' to analysis command (available only with -o option)
 -e [k=v]                    set an evaluable config variable (available only with -o option)
 -c [output.pac] [input.sig] parses a FLIRT signature and converts it to its other format
 -o [output.sig] [input.bin] performs an analysis on the binary and generates the FLIRT signature.
 -d [flirt.sig]              parses a FLIRT signature and dump its content
 -q                          quiet mode
 -v                          show version information
Examples:
  rz-sign -d signature.sig
  rz-sign -c new_signature.pac old_signature.sig
  rz-sign -o libc.sig libc.so.6
```

**Test plan**

Generation of valid files which can be parsed and applied to rizin and other tools.

How to test the PR:
Open a file (possibly with symbols) run `aa ; zfc /path/to/my/new/flirt.sig` then strip the file and reopen it again, then `aaa ; zfs /path/to/my/new/flirt.sig`

**Closing issues**

Closes #817

XVilka commented on librz/core/cconfig.c:
```diff
@@ -2892,6 +2892,15 @@ static bool cb_dbg_verbose(void *user, void *data) {
 	return true;
 }
 
+static bool cb_flirt(void *user, void *data) {
```
Add `rz_return_val_if_fail()`

XVilka commented on librz/core/cfile.c:
```diff
@@ -888,25 +889,39 @@ static bool map_multi_dex(RzCore *core, RzIODesc *desc, ut32 id) {
+	RzBinFile *cur = rz_pvector_empty(&cf->binfiles) ? NULL : rz_pvector_tail(&cf->binfiles);
+	if (cur) {
+		RzIODesc *iod = rz_io_desc_get(core->io, cur->fd);
+		if (iod) {
```
Maybe reverse the condition and switch to the "early exit" pattern?

wargio replied:
This condition cannot be reversed because it is an optional one.

XVilka commented on librz/core/cfile.c:
```diff
@@ -888,25 +889,39 @@ static bool map_multi_dex(RzCore *core, RzIODesc *desc, ut32 id) {
+	RzCoreFile *cf = rz_core_file_cur(core);
+
+	// adds the current size and align next address
+	ut64 base_address = 0x200000;
```
Make that a define maybe?

XVilka commented on librz/core/cfile.c:
```diff
@@ -1098,14 +1113,18 @@ RZ_API bool rz_core_bin_load(RZ_NONNULL RzCore *r, RZ_NULLABLE const char *filen
 	}
 
+	if (!base_address) {
+		base_address = 0x08000000;
```
Same, move that to a define, I think.

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,249 @@
+
+	rz_list_foreach (tokens, it, tk) {
+		for (ut32 i = 0; i < RZ_ARRAY_SIZE(file_map); ++i) {
+			if (!strcmp(tk, file_map[i].name)) {
```
Reverse the condition to call "continue"

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,249 @@
+
+	rz_list_foreach (tokens, it, tk) {
+		for (ut32 i = 0; i < RZ_ARRAY_SIZE(os_map); ++i) {
+			if (!strcmp(tk, os_map[i].name)) {
```
Same, reverse the condition maybe?

XVilka commented on librz/signature/flirt.c:
```diff
@@ -1158,6 +935,41 @@ static int parse_v10_header(RzBuffer *buf, idasig_v10_t *header) {
+
+exit:
+	free(header);
+
```
Unnecessary empty lines.

XVilka commented:
```
FAILED: librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o 
tcc -Ilibrz/signature/librz_sign.so.0.4.0-git.p -I. -I.. -I../librz/include -Isubprojects/sdb -I../subprojects/sdb -Isubprojects/sdb/src -I../subprojects/sdb/src -I../librz/analysis/arch/gb -I../librz/type/parser -Wall -O2 -g -D_GNU_SOURCE -Werror=sizeof-pointer-memaccess -fvisibility=hidden -DRZ_PLUGIN_INCORE=1 -MD -MF librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o.d -o librz/signature/librz_sign.so.0.4.0-git.p/flirt.c.o -c ../librz/signature/flirt.c
../librz/signature/flirt.c:670: error: function-like macro 'if' is not defined
```

wargio committed:
Refactoring rz_analysis_mask definition

wargio committed:
[FLIRT] Add pac and sig creation with options & compression

public functions and references are generated in a sequential way besides
the first one which is the first to be assigned.
since we do not search for sequential references, then we can avoid to
add extra public or referenced functions

wargio committed:
Added rz_str_startswith_const and rz_list_iter_get_next_data

wargio committed:
[FLIRT] Rewritten rz-sign

wargio committed:
Fix tests

wargio commented:
ASAN here: https://github.com/rizinorg/rizin/pull/2079

wargio committed:
Add doxygen to rz_list_*

wargio committed:
[FLIRT] Add missing arch (arc, pic, riscv, spc700 xtensa, and more)

wargio committed:
Add doxygen to rz_strbuf_slice

wargio commented:
maybe the sig options should be under `flirt.sig.*`

XVilka commented:
@wargio I think better the opposite: `sig.flirt.*`

wargio commented:
> @wargio I think better the opposite: `sig.flirt.*`

i'm referring to the file type options, so `flirt.sig.*` and `flirt.pac.*`

XVilka commented on librz/core/cconfig.c:
```diff
@@ -3787,6 +3797,24 @@ RZ_API int rz_core_config_init(RzCore *core) {
+		"pe", "pilot", "srec", "w32run", "zip", "all", "none", NULL);
+	n = NODECB("flirt.header.app", "all", &cb_flirt);
+	SETDESC(n, "FLIRT app list (comma separated) for sig format");
+	SETOPTIONS(n, "console", "graphics", "exe", "dll", "drv", "thread:single", "thread:multi", "16bit", "32bit", "64bit", "all", "none", NULL);
```
Shouldn't this option be inferred automatically?

wargio replied:
tbh, we do not use and setting it to all is ok

XVilka replied:
Yes, I think we should remove it.

wargio replied:
you mean, remove the option? i think this option is needed only if you use IDA and anyway setting it to FFFF (i.e. all on rizin) like i have seen in all the other signatures is fine.

XVilka replied:
Yes

wargio replied:
removed.

XVilka commented on librz/core/cmd_descs/cmd_zign.yaml:
```diff
@@ -185,6 +185,12 @@ commands:
     subcommands:
+      - name: zfc
+        cname: zign_flirt_create
+        summary: Create FLIRT a file
```
Probably better to reword. Completely unclear

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,266 @@
+	{ "pic", RZ_FLIRT_SIG_ARCH_PIC16 },
+	{ "riscv", RZ_FLIRT_SIG_ARCH_RISCV },
+	// { "rl78", RZ_FLIRT_SIG_ARCH_RL78 },
+	// { "s390", RZ_FLIRT_SIG_ARCH_S390 },
```
"sysz" in Rizin.

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,266 @@
+	{ "cr16", RZ_FLIRT_SIG_ARCH_CR16 },
+	// { "mn102l00", RZ_FLIRT_SIG_ARCH_MN102L00 },
+	// { "tms320c1x", RZ_FLIRT_SIG_ARCH_TMS320C1X },
+	// { "nec_v850x", RZ_FLIRT_SIG_ARCH_NEC_V850X },
```
"v850" is also supported.

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,266 @@
+	// { "i960", RZ_FLIRT_SIG_ARCH_I960 },
+	// { "f2mc", RZ_FLIRT_SIG_ARCH_F2MC },
+	// { "tms320c54", RZ_FLIRT_SIG_ARCH_TMS320C54 },
+	// { "tms320c55", RZ_FLIRT_SIG_ARCH_TMS320C55 },
```
Supported by Rizin as well

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,266 @@
+	// { "tms", RZ_FLIRT_SIG_ARCH_TMS },
+	{ "6502", RZ_FLIRT_SIG_ARCH_6502 },
+	// { "pdp", RZ_FLIRT_SIG_ARCH_PDP },
+	// { "68k", RZ_FLIRT_SIG_ARCH_68K },
```
Supported as well

XVilka commented on librz/core/csign.c:
```diff
@@ -7,6 +7,266 @@
+	{ "x86", RZ_FLIRT_SIG_ARCH_386 },
+	{ "z80", RZ_FLIRT_SIG_ARCH_Z80 },
+	// { "i860", RZ_FLIRT_SIG_ARCH_I860 },
+	// { "8051", RZ_FLIRT_SIG_ARCH_8051 },
```
Also supported

XVilka commented on librz/signature/create.c:
```diff
@@ -0,0 +1,458 @@
+extern ut16 flirt_crc16(const ut8 *data_p, size_t length);
+
+#if 0
+static void sig_dbg_node(const char *name, const RzFlirtNode *node) {
```
Maybe remove this?

XVilka commented on librz/util/list.c:
```diff
@@ -18,6 +18,15 @@ RZ_API RzListIter *rz_list_iter_get_next(RzListIter *list) {
 	return list->n;
 }
 
+RZ_API void *rz_list_iter_get_next_data(RzListIter *list) {
```
Add Doxygen please

XVilka commented on librz/util/strbuf.c:
```diff
@@ -107,13 +107,18 @@ RZ_API bool rz_strbuf_setbin(RzStrBuf *sb, const ut8 *s, size_t l) {
 }
 
-// TODO: there's room for optimizations here
 RZ_API bool rz_strbuf_slice(RzStrBuf *sb, int from, int len) {
```
Doxygen.

XVilka commented on test/db/cmd/cmd_zignature:
```diff
@@ -1794,7 +1794,7 @@ aaa
 EOF
 EXPECT=<<EOF
-Found 207 FLIRT signatures via bins/flirt/elf-x86/libcurl.a.pac
+Found 100 FLIRT signatures via bins/flirt/elf-x86/libcurl.a.pac
```
Why such a big difference?

wargio replied:
because i was parsing the pac file in the wrong way

XVilka commented on test/db/cmd/cmd_zignature:
```diff
@@ -1815,6 +1815,6 @@ CMDS=<<EOF
 zfd bins/flirt/elf-x86/libcurl.a.pac~?
 EOF
 EXPECT=<<EOF
-166
```
Same

wargio replied:
same issue

XVilka reviewed (changes requested):
Overall is quite good. Awesome work! 💯

wargio committed:
[FLIRT] renamed flirt.* and rz-sign output when not in quiet

XVilka commented on librz/core/cmd_descs/cmd_zign.yaml:
```diff
@@ -185,15 +185,21 @@ commands:
       - name: zfd
         cname: zign_flirt_dump
-        summary: Open FLIRT file and dump
+        summary: Opens a FLIRT file (.pac or .sigà) and dumps its contents
```
Before you used "create" now it "opens", please use either. I think it's better to keep "open"

XVilka commented on librz/core/cmd_descs/cmd_zign.yaml:
```diff
@@ -185,15 +185,21 @@ commands:
       - name: zfs
         cname: zign_flirt_scan
-        summary: Open FLIRT file and scan
+        summary: Opens a FLIRT file and tries to apply the signatures to the loaded binary
```
Same

wargio committed:
Fix zfc,zfd and zfs command description

wargio committed:
Removed flirt.sig.app option.

wargio commented:
the question is now, should we squash it?

XVilka commented:
Yes, squashing is fine.

ret2libc commented:
@wargio could you please provide more detailed "test instructions" in the first comment of the PR? Since there are a lot of changes i'm having an hard time understanding what i should check.

wargio commented:
I have updated the test instructions

ret2libc commented:
```
Indirect leak of 288 byte(s) in 12 object(s) allocated from:                                   
    #0 0x7f238dc2193f in __interceptor_malloc (/lib64/libasan.so.6+0xae93f)                    
    #1 0x7f238d38b34d in rz_list_append ../librz/util/list.c:293                               
    #2 0x7f237f743945 in rz_sign_flirt_node_new ../librz/signature/create.c:426                
    #3 0x7f237dfa8a43 in rz_core_flirt_create_file ../librz/core/csign.c:420                   
    #4 0x7f237e15dd68 in rz_zign_flirt_create_handler ../librz/core/cmd/cmd_zign.c:1423        
```

ret2libc commented:
```
Direct leak of 32 byte(s) in 1 object(s) allocated from:                                       
    #0 0x7f238dc21af7 in calloc (/lib64/libasan.so.6+0xaeaf7)                                  
    #1 0x7f238d38b0a9 in rz_list_new ../librz/util/list.c:236                                  
    #2 0x7f238d38b0de in rz_list_newf ../librz/util/list.c:249                                 
    #3 0x7f237f742e73 in rz_sign_flirt_node_new ../librz/signature/create.c:368                
    #4 0x7f237dfa8a43 in rz_core_flirt_create_file ../librz/core/csign.c:420                   
    #5 0x7f237e15dd68 in rz_zign_flirt_create_handler ../librz/core/cmd/cmd_zign.c:1423        

```

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1102,7 +1103,14 @@ RZ_API RzCoreAutocomplete *rz_core_autocomplete_find(RzCoreAutocomplete *parent,
 RZ_API void rz_core_analysis_propagate_noreturn(RzCore *core, ut64 addr);
 
-RZ_API void rz_core_flirt_dump(RZ_NONNULL const char *flirt_file);
+RZ_API void rz_core_flirt_dump_node(RZ_NONNULL const RzFlirtNode *node);
```
is really necessary to export this function?

ret2libc commented on test/db/formats/ar:
```diff
@@ -129,150 +129,150 @@ nth paddr      vaddr      bind   type   size lib name
 nth paddr      vaddr      bind   type size lib name             
 ----------------------------------------------------------------
 1   ---------- 0x00000000 LOCAL  FILE 0        arch.c
-2   0x00000040 0x00000040 LOCAL  SECT 0        .text
```
why these changes? What do they have to do with flirt?

wargio replied:
needed to properly map all the .o files inside an .a

ret2libc replied:
how is that related to flirt?

wargio replied:
ok, let me drop the commit on multibin and open them in a different PR

wargio committed:
Resolved comments and fix memleak

wargio committed:
Revert test linked to dropped commit

ret2libc commented on librz/util/list.c:
```diff
@@ -13,16 +13,41 @@ RZ_API void rz_list_iter_free(RzListIter *list) {
+/**
+ * \brief returns the value stored in the next RzList iterator
+ *
+ * */
```
```suggestion
 **/
```

wargio replied:
i think that is added by clang-format.

ret2libc replied:
weird, unless they have changed something very recently. I think it should be `/** **/`

ret2libc commented on librz/include/rz_util/rz_str.h:
```diff
@@ -237,6 +237,7 @@ RZ_API bool rz_str_glob(const char *str, const char *glob);
 RZ_API int rz_str_binstr2bin(const char *str, ut8 *out, int outlen);
 RZ_API char *rz_str_between(const char *str, const char *prefix, const char *suffix);
 RZ_API bool rz_str_startswith(const char *str, const char *needle);
+#define rz_str_startswith_const(b, const_str) (!strncmp(b, const_str, strlen(const_str)))
```
I think this can be avoided and we would have the same effect by moving the rz_str_startswith definition in the include file with `static inline`. That way *you*, as developer, don't have to know whether the string may or may not be const and we don't have to duplicate essentially the same code as `rz_str_startswith` here.

wargio replied:
let me remove that.

ret2libc reviewed (commented):
Please fix just these two comments, then LGTM! Sorry for the delay

wargio committed:
Fix comments
